Feature filter keep_keys.

vbotka opened the pull request:
##### SUMMARY

ansible.utils.keep_keys does not work properly. Fixing is not efficient.

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
community.general.keep_keys

##### ADDITIONAL INFORMATION
This PR implements #8438

vbotka commented:
I see the tag `has_issue`. What issue does this PR have?

felixfontein commented:
> I see the tag `has_issue`. What issue does this PR have?

Issue #8438 references this PR, so the bot adds `has_issue` to indicate that an issue references this PR.

(You probably want to add something like `Fixes #8438` to the PR's description so GitHub will auto-close the issue once the PR is merged.)

vbotka commented:
> ... add something like `Fixes #8438` to the PR's description so GitHub will auto-close the issue once the PR is merged.

Thank you! I updated the description.

felixfontein commented on changelogs/fragments/8439-feature-keep_keys.yml:
```diff
@@ -0,0 +1,2 @@
+major_changes:
+  - New filter keep_keys. https://github.com/ansible-collections/community.general/pull/8439
```
Please remove the changelog fragment; new plugins are automatically announced in the changelog due to their top-level `version_added` in the documentation.

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+DOCUMENTATION = '''
+    name: keep_keys
+    short_description: Keep specific keys from dictionaries in a list.
+    version_added: "2.17"
```
```suggestion
    version_added: "9.1.0"
```

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+
+DOCUMENTATION = '''
+    name: keep_keys
+    short_description: Keep specific keys from dictionaries in a list.
```
`short_description` is the only description that shouldn't have a full sentence :)
```suggestion
    short_description: Keep specific keys from dictionaries in a list
```

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+      target:
+        description:
+          - A list of keys or keys patterns to keep.
+          - The interpretation of the list items depends on the option C(matching_parameter)
```
```suggestion
          - The interpretation of the list items depends on the option O(matching_parameter)
```

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+        description:
+          - A list of keys or keys patterns to keep.
+          - The interpretation of the list items depends on the option C(matching_parameter)
+          - For matching_parameter C(regex) only the first item is taken.
```
```suggestion
          - For O(matching_parameter=regex) only the first item is taken.
```

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+        type: str
+        default: equal
+        choices:
+          - equal
+          - starts_with
+          - ends_with
+          - regex
```
```suggestion
          equal: Matches keys of exactly one of the names in O(target).
          starts_with: Matches keys that start with one of the names in O(target).
          ends_with: Matches keys that end with one of the names in O(target).
          regex:
            - Matches keys that match the regular expresion provided in O(target).
            - In this case, O(target) must have exactly one entry.
```

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+        if isinstance(target, string_types):
+            tt = target
+        elif isinstance(target, Sequence):
+            tt = target[0]
```
1. You need to check whether the sequence has more than zero elements.
2. I would fail in case there are more than one elements. It's better to fail in such cases than to let the user wonder for an hour why it only uses the first regex, until they notice part of the documentation that says that the others are ignored.

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+            raise AnsibleFilterError(msg % target)
+        try:
+            re.compile(tt)
+            is_valid = True
+        except re.error:
+            is_valid = False
+        if not is_valid:
+            msg = ("The target for keep_keys must be a valid regex if matching_parameter is regex."
+                   "target is %s")
+            raise AnsibleFilterError(msg % tt)
```
Why not simplify this to:
```suggestion
        except re.error:
            msg = ("The target for keep_keys must be a valid regex if matching_parameter is regex."
                   "target is %s")
            raise AnsibleFilterError(msg % tt)
```

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+                   "target is %s")
+            raise AnsibleFilterError(msg % tt)
+    else:
+        if not isinstance(tt, Sequence):
```
Strings are also sequences, so better handle that case first:
```suggestion
        if isinstance(tt, string_types):
            tt = [tt]
        if not isinstance(tt, Sequence):
```
Otherwise if the user passes `target='foo'`, it will treat it as a three-element list `['f', 'o', 'o']`.

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+            tt = target
+        else:
+            tt = target[0]
+        my_keys = [[k for k in i.keys() if re.match(tt, k)] for i in ld]
```
In case of `mp == 'regex'`, you already compiled the regular expression. But now you're using `re.match()` to match the regular expression *as a string* against a large set of strings. You should really use the compilation result here.

(The current version is likely still efficient because `re` does some magic to cache recently used regular expressions, but simply relying on that is not a good idea, and also still less efficient than simply using the compiled version you already could have here.)

felixfontein commented on plugins/filter/keep_keys.py:
```diff
@@ -0,0 +1,154 @@
+            tt = target[0]
+        my_keys = [[k for k in i.keys() if re.match(tt, k)] for i in ld]
+
+    return [dict([(k, ld[i][k]) for k in j]) for i, j in enumerate(my_keys)]
```
I personally find the way this plugin works very confusing.

Why not simply create a different predicate `keep_key(key)` in an `if` based on `mp`, like this:
```py
    if mp == 'equal':
        def keep_key(key):
            return key in tt
    elif mp == 'starts_with':
        tt_tuple = tuple(tt)
        def keep_key(key):
            return key.starts_with(tt_tuple)
    elif mp == 'ends_with':
        tt_tuple = tuple(tt)
        def keep_key(key):
            return key.ends_with(tt_tuple)
    elif mp == 'regex':
        tt_compile = ... (compilation result from above)
        def keep_key(key):
            return tt_compile.match(key) is not None
```
Then you can simply compute the result as:
```py
    return [dict((k, v) for k, v in d.items() if keep_key(k)) for d in data]
```

felixfontein reviewed (commented):
Thanks for your contribution!

One general problem this has is that it assume that all dictionary keys are strings. That's generally not the case. (It is true if the input is JSON, but since this is a filter it can also have totally different input.) The code needs to be adjusted to handle non-string keys in some consistent way (either throw them all out, or keep them all, I guess).

(WIth the predicate functions - see below - this should be pretty easy to do.)

vbotka commented:
Thank you very much! I agree with all the points. Very useful. I'll create a new PR.

felixfontein commented:
You can also update the same PR, whatever you prefer :)

vbotka commented:
I created a new PR #8456. I implemented all your proposals, I hope. I find your contribution substantial. May I add you as the author to the source?

Notes on selected topics:

* The fragmentation of plugins/plugin_utils/keys_filter.py

* _keys_filter_params test parameters:

    * data must be a list of dictionaries. All keys must be strings.
    * target must be a non-empty sequence.
    * matching_parameter is member of a list.

* keys_filter_target_str test:

    * target is a list of strings, or
    * target is a string or list with single string if matching_parameter=regex
    
convert, and return:

    * tuple of unique target items, or
    * compiled regex if matching_parameter=regex

* keys_filter_target_dict is prepared for the filter replace_keys

* O(target) can be string for O(matching_parameter=regex). I find closing regex in a list counterintuitive.
* Therefore, I changed the type of O(target) to raw.

felixfontein commented:
> May I add you as the author to the source?

If you wish so, sure :) Thanks!

vbotka commented:
Let's close this one and proceed in #8456. Thank you!
